app/javascript/mastodon/components/status_list.js

```diff
@@ -53,6 +60,39 @@ class StatusList extends ImmutablePureComponent {
 
   componentWillUnmount () {
     this.detachScrollListener();
+    this.detachIntersectionObserver();
+  }
+
+  attachIntersectionObserver () {
+    const onIntersection = (entries) => {
+      const isIntersecting = { };
+
+      entries.forEach(entry => {
+        const statusId = entry.target.getAttribute('data-id');
+
+        isIntersecting[statusId] = entry.isIntersecting;
+      });
+
+      this.setState(state => ({
+        isIntersecting: Object.assign({ }, state.isIntersecting, isIntersecting),
+      }));
+    };
+
+    const options = {
+      root: this.node,
+      rootMargin: '100% 0px',
+    };
```

Comment:
I think we may be able to get better performance by using something like `500%` instead of `100%`. As is, this re-renders components almost constantly as you scroll, leading to very jerky scrolling because of the bottleneck you describe. Whereas if we allowed for 5x the available viewport height, then that should lead to less jumpy scrolling because it won't start un-rendering until you've scrolled 5x of the way down. I admit I still need to test this theory on a mobile device, though.